### ISSUE TITLE
_document  -> makeStylesheetInert -> "Cannot read property 'type' of null" if Child is null

### ISSUE THREAD

**Issue by pecoram** [vercel/next.js#24783]
### What version of Next.js are you using?

10.2.0

### What version of Node.js are you using?

14.16.0

### What browser are you using?

All

### What operating system are you using?

Windows

### How are you deploying your application?

Other platform

### Describe the Bug

When a Child of the Head is null an exception is thrown: `TypeError: Cannot read property 'type' of null`.

In the condition a nullsafe is missing

https://github.com/vercel/next.js/blob/cf4ba8d7051a4fc47647b7c7bc3b7f4b45de41eb/packages/next/pages/_document.tsx#L331

### Expected Behavior

```
if (c && c.type === 'link' &&
        c.props['href'] &&
        OPTIMIZED_FONT_PROVIDERS.some((url) => c.props['href'].startsWith(url))
      ) 
```

### To Reproduce

In the _document put a null component inside the Head.

```
import Document, { Html, Head, Main, NextScript } from 'next/document'

const NullComponent = () => null;

class MyDocument extends Document {
  static async getInitialProps(ctx) {
    const initialProps = await Document.getInitialProps(ctx)
    return { ...initialProps }
  }

  render() {
    return (
      <Html>
        <Head>
            <NullComponent />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

export default MyDocument
```

**Comment by ljosberinn** [comment]
Can you elaborate why you're rendering a NullComponent in Head? Head should contain regular head-elements such as link, script, meta, not nothing.

**Comment by pecoram** [comment]
for example when you write something like that:

`{ condition && <link rel="preload" as="image" href="" /> }`

if condition is false you get the error!

**Comment by pecoram** [comment]
now I solved it this way:

`{condition ? <Component /> : ('')}`

but a nullsafe in that point is a better and safe solution!

**Comment by ljosberinn** [comment]
agreed, I'll try to PR this later today

**Comment by synoptase** [comment]
I'm having a somewhat similar issue:  `TypeError: Cannot assign to read only property 'children' of object '#<Object>'`
Traced back to: https://github.com/vercel/next.js/blob/cf4ba8d7051a4fc47647b7c7bc3b7f4b45de41eb/packages/next/pages/_document.tsx#L340

**nextjs** 10.2.0
**node** 12.22.1

Still trying to figure out the root cause though.

**Comment by ljosberinn** [comment]
@pecoram could you try to create a reproducible repo?

Both
```js
import Head from "next/head";

export default function Test() {
  return (
    <>
      <Head>{false && <link rel="preload" as="image" href="" />}</Head>
      <h1>hi</h1>
    </>
  );
}

```
and 
```js
import Head from "next/head";

export default function Test() {
  return (
    <>
      <Head>{null}</Head>
      <h1>hi</h1>
    </>
  );
}

```
work flawlessly for me

@synoptase same for you please, 
```js
import Head from "next/head";

export default function Test() {
  return (
    <>
      <Head>
        {/* // eslint-disable-next-line react/jsx-no-useless-fragment */}
        <>{null}</>
      </Head>
      <h1>hi</h1>
    </>
  );
}

```  works fine too

**Comment by synoptase** [comment]
@ljosberinn it worked indeed in my dev env. But once build and deployed, error kept appearing until i removed `Fragment`. It definitely seems related to the build env, but i just tried reproducing the issue by building and running my docker image locally, without any luck... Not a bug for me initial comment then.

**Comment by timneutkens** [comment]
This is a bug, it's related to the new font optimization (which only runs during build).

**Comment by ljosberinn** [comment]
I believe changing the implementation of `_document.makeStylesheetInert` to this should fix it but haven't found a way to test it integratively:

```ts
  makeStylesheetInert(node: ReactNode): ReactNode[] {
    return React.Children.map(node, (child) => {
      if (!React.isValidElement(child)) {
        return null
      }

      if (
        child.type === 'link' &&
        child.props.href &&
        OPTIMIZED_FONT_PROVIDERS.some((url) => child.props.href.startsWith(url))
      ) {
        return React.cloneElement(child, {
          ...child.props,
          'data-href': child.props.href,
          href: undefined,
        })
      }

      if (child.props?.children) {
        return React.cloneElement(child, {
          ...child.props,
          children: this.makeStylesheetInert(child.props.children),
        })
      }

      return child
    })
  }
  ```

**Comment by fbudassi** [comment]
I'm also experiencing a similar problem:
Node.js: 14.16.0
Next.js: 10.2.0

Error message:
`TypeError: Cannot assign to read only property 'children' of object '#<Object>'`

But in my case, this is the line inside the `_document.tsx`'s Head that triggers the problem:

```
<style>{`#__next { height: 100%; }`}</style>
```
It wasn't failing in Next.js 10.0.6.

**Comment by janicklas-ralph** [comment]
I'm having a hard time reproducing this bug. We also already [check for this edge case](https://github.com/vercel/next.js/blob/canary/test/integration/font-optimization/fixtures/with-google/pages/_document.js#L22)

Could you make a small example where the error is reproducible?

**Comment by flybayer** [comment]
> Error message:
> `TypeError: Cannot assign to read only property 'children' of object '#<Object>'`
> 

Maybe there's two bugs here, but I get this error message with `NODE_ENV=test next build` which is this bug: https://github.com/vercel/next.js/issues/25491

**Comment by atdrago** [comment]
Our project suddenly started returning 500s in production because of `TypeError: Cannot assign to read only property 'children' of object '#<Object>'`. 

I've created a project where I've reproduced the issue (https://github.com/atdrago/repro-24783). Here's the commit where the issue is introduced: https://github.com/atdrago/repro-24783/commit/1ea71f6e8bf8c420b8e0b521a30b3ef6d6592371

**Steps to reproduce:**
1. `npm run build && npm start`
2. Navigate to `http://localhost:3000/foo/bar` (`/foo/bar` isn't special, it can be `/foo/anything-here`)
3. Observe the 500 error

**Some things I noticed:**
- It is only reproducible using `next-boost`
- It is only reproducible using `getStaticPaths` and navigating to a route that is not caught by the `paths` array, and does not 404
- It is only reproducible if `getStaticPaths` returns a `fallback` value of `true` or `'blocking'` (since `false` would 404)
